### PR TITLE
Add newline to text dataset builder for controlling universal newlines mode

### LINKED ISSUE #3804: Text builder with custom separator line boundaries
**Is your feature request related to a problem? Please describe.**
The current [Text](https://github.com/huggingface/datasets/blob/207be676bffe9d164740a41a883af6125edef135/src/datasets/packaged_modules/text/text.py#L23) builder implementation splits texts with `splitlines()` which splits the text on several line boundaries. Not all of them are always wanted. 

**Describe the solution you'd like**
```python
                if self.config.sample_by == "line":
                    batch_idx = 0
                    while True:
                        batch = f.read(self.config.chunksize)
                        if not batch:
                            break
                        batch += f.readline()  # finish current line
                        if self.config.custom_newline is None: 
                            batch = batch.splitlines(keepends=self.config.keep_linebreaks)
                        else:
                            batch = batch.split(self.config.custom_newline)[:-1]
                        pa_table = pa.Table.from_arrays([pa.array(batch)], schema=schema)
                        # Uncomment for debugging (will print the Arrow table size and elements)
                        # logger.warning(f"pa_table: {pa_table} num rows: {pa_table.num_rows}")
                        # logger.warning('\n'.join(str(pa_table.slice(i, 1).to_pydict()) for i in range(pa_table.num_rows)))
                        yield (file_idx, batch_idx), pa_table
                        batch_idx += 1
```

**A clear and concise description of what you want to happen.**
Creating the dataset rows with a subset of the `splitlines()` line boundaries.

### PR DESCRIPTION
**Pull request by albertvillanova**
Fix #3804.

### CODE REVIEW

**Comment by albertvillanova** [comment]
The failing CI test has nothing to do with this PR.

File: src/datasets/packaged_modules/text/text.py [hunk 1]
```diff
@@ -53,43 +53,55 @@ def _split_generators(self, dl_manager):
...
+                        batch += new_batch
                         if not batch:
                             break
-                        batch += f.readline()  # finish current line
-                        # StringIO.readlines, by default splits only on "\n" (and keeps line breaks)
-                        batch = StringIO(batch).readlines()
-                        if not self.config.keep_linebreaks:
-                            batch = [line.rstrip("\n") for line in batch]
-                        pa_table = pa.Table.from_arrays([pa.array(batch)], schema=schema)
+                        batch = batch.split(self.config.newline or "\n")
```
Comment (lhoestq) [inline src/datasets/packaged_modules/text/text.py:68]: If the batch ends on "\n", here you will end up with `'...\n'.split("\n")` which is equal to `['...', '']` which is maybe not expected

**Review by lhoestq — COMMENTED** [review]
Thanks and sorry for the delay, I added some comments:

**Comment by albertvillanova** [comment]
I'm closing this PR.

**Pull request closed without merging**